File: packages/shared-types/src/fluid-balance.ts
/** Fluid Balance (I/O-Bilanz) types. */

// ─── Direction & Category Types ────────────────────────────────

export type FluidDirection = "intake" | "output";

export type IntakeCategory = "infusion" | "oral" | "medication" | "blood_product" | "nutrition" | "other_intake";
export type OutputCategory = "urine" | "drain" | "vomit" | "stool" | "perspiratio" | "blood_loss" | "other_output";
export type FluidCategory = IntakeCategory | OutputCategory;

export type FluidRoute = "iv" | "oral" | "subcutaneous" | "rectal" | "ng_tube" | "catheter" | "other";

// ─── Labels ────────────────────────────────────────────────────

export const FLUID_DIRECTION_LABELS: Record<FluidDirection, string> = {
    intake: "Einfuhr",
    output: "Ausfuhr",
};

export const FLUID_CATEGORY_LABELS: Record<string, string> = {
    // Intake
    infusion: "Infusion",
    oral: "Oral",
    medication: "Medikamentös",
    blood_product: "Blutprodukt",
    nutrition: "Enterale Ernährung",
    other_intake: "Andere Einfuhr",
    // Output
    urine: "Urin",
    drain: "Drainage",
    vomit: "Erbrechen",
    stool: "Stuhl",
    perspiratio: "Perspiratio insensibilis",
    blood_loss: "Blutverlust",
    other_output: "Andere Ausfuhr",
};

export const INTAKE_CATEGORIES: IntakeCategory[] = [
    "infusion", "oral", "medication", "blood_product", "nutrition", "other_intake",
];

export const OUTPUT_CATEGORIES: OutputCategory[] = [
    "urine", "drain", "vomit", "stool", "perspiratio", "blood_loss", "other_output",
];

export const FLUID_ROUTE_LABELS: Record<FluidRoute, string> = {
    iv: "Intravenös",
    oral: "Oral",
    subcutaneous: "Subkutan",
    rectal: "Rektal",
    ng_tube: "Magensonde",
    catheter: "Katheter",
    other: "Andere",
};

// ─── Interfaces ────────────────────────────────────────────────

export interface FluidEntry {
    id: string;
    patient_id: string;
    encounter_id?: string;
    direction: FluidDirection;
    category: string;
    display_name: string;
    volume_ml: number;
    route?: FluidRoute;
    recorded_at: string;
    recorded_by?: string;
    notes?: string;
    created_at: string;
}

export interface FluidEntryCreate {
    patient_id: string;
    encounter_id?: string;
    direction: FluidDirection;
    category: string;
    display_name: string;
    volume_ml: number;
    route?: string;
    recorded_at?: string;
    notes?: string;
}

export interface FluidEntryUpdate {
    volume_ml?: number;
    display_name?: string;
    route?: string;
    notes?: string;
}

export interface FluidBalanceSummary {
    period_start: string;
    period_end: string;
    total_intake_ml: number;
    total_output_ml: number;
    balance_ml: number;
    intake_by_category: Record<string, number>;
    output_by_category: Record<string, number>;
    entry_count: number;
}

export interface FluidBalanceMeta {
    direction_labels: Record<string, string>;
    category_labels: Record<string, string>;
    intake_categories: string[];
    output_categories: string[];
    route_labels: Record<string, string>;
}

File: backend/src/domain/schemas/fluid_balance.py
"""Fluid balance (I/O-Bilanz) Pydantic schemas."""

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

# ─── Category Constants ─────────────────────────────────────────

INTAKE_CATEGORIES = ("infusion", "oral", "medication", "blood_product", "nutrition", "other_intake")
OUTPUT_CATEGORIES = ("urine", "drain", "vomit", "stool", "perspiratio", "blood_loss", "other_output")
ALL_CATEGORIES = INTAKE_CATEGORIES + OUTPUT_CATEGORIES

DIRECTION_LABELS: dict[str, str] = {
    "intake": "Einfuhr",
    "output": "Ausfuhr",
}

CATEGORY_LABELS: dict[str, str] = {
    # Intake
    "infusion": "Infusion",
    "oral": "Oral",
    "medication": "Medikamentös",
    "blood_product": "Blutprodukt",
    "nutrition": "Enterale Ernährung",
    "other_intake": "Andere Einfuhr",
    # Output
    "urine": "Urin",
    "drain": "Drainage",
    "vomit": "Erbrechen",
    "stool": "Stuhl",
    "perspiratio": "Perspiratio insensibilis",
    "blood_loss": "Blutverlust",
    "other_output": "Andere Ausfuhr",
}

ROUTE_LABELS: dict[str, str] = {
    "iv": "Intravenös",
    "oral": "Oral",
    "subcutaneous": "Subkutan",
    "rectal": "Rektal",
    "ng_tube": "Magensonde",
    "catheter": "Katheter",
    "other": "Andere",
}


# ─── Schemas ────────────────────────────────────────────────────

class FluidEntryCreate(BaseModel):
    patient_id: uuid.UUID
    encounter_id: uuid.UUID | None = None
    direction: str = Field(..., pattern=r"^(intake|output)$")
    category: str = Field(..., min_length=1, max_length=30)
    display_name: str = Field(..., min_length=1, max_length=200)
    volume_ml: float = Field(..., gt=0)
    route: str | None = None
    recorded_at: datetime | None = None
    notes: str | None = None


class FluidEntryUpdate(BaseModel):
    volume_ml: float | None = Field(None, gt=0)
    display_name: str | None = None
    route: str | None = None
    notes: str | None = None


class FluidEntryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    patient_id: uuid.UUID
    encounter_id: uuid.UUID | None
    direction: str
    category: str
    display_name: str
    volume_ml: float
    route: str | None
    recorded_at: datetime
    recorded_by: uuid.UUID | None
    notes: str | None
    created_at: datetime


class PaginatedFluidEntries(BaseModel):
    items: list[FluidEntryResponse]
    total: int
    page: int = 1
    per_page: int = 50


class FluidBalanceSummary(BaseModel):
    """24-hour fluid balance summary."""
    period_start: datetime
    period_end: datetime
    total_intake_ml: float
    total_output_ml: float
    balance_ml: float  # intake - output (positive = net gain)
    intake_by_category: dict[str, float]  # category -> total mL
    output_by_category: dict[str, float]
    entry_count: int
